### ISSUE TITLE
orphan blocks - optimization

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#521]
Right now we maintain a pool of up to 50 orphan blocks.
Every time we process a valid block we brute force check _all_ the orphans to see if any are now valid and can be accepted on the chain.

We see a _lot_ of activity during sync where we are repeatedly checking large numbers of orphans over and over as we process individual blocks.

Proposal -
1) store orphan blocks in a HashMap keyed by `prev_hash`
    * additional index by `hash` (we need both to be fast lookups)
2) if we process a block, check the orphan pool for -
    * orphan indexed by `prev_hash` that matches the hash of the block we just processed
    * process it
3) repeat

Additional -
1) think through expiration of orphans in the orphan pool
    - do we want these to hang around forever?
     - which are more important and how do we sort them?

**Comment by antiochp** [comment]
For example we see this in between successfully processing a couple of blocks - 

```
Dec 21 09:26:48.318 DEBG pipe: process_block 24f7d1c6 at 23719 with 2 inputs and 5 outputs.
Dec 21 09:26:48.318 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.324 DEBG Starting new sumtree extension.
Dec 21 09:26:48.531 DEBG pipe: proces_block 24f7d1c6 at 23719 is valid, save and append.
Dec 21 09:26:48.533 INFO pipe: update_head: 24f7d1c6, 177870939 at 23719
Dec 21 09:26:48.533 DEBG Committing sumtree extension.
Dec 21 09:26:48.788 DEBG Sumtree extension done.
Dec 21 09:26:48.788 DEBG check_orphans: # orphans 45
Dec 21 09:26:48.789 DEBG pipe: process_block 1fec2cfe at 23729 with 2 inputs and 5 outputs.
Dec 21 09:26:48.789 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.792 DEBG process_block: orphan: 1fec2cfe, # orphans 45
Dec 21 09:26:48.793 DEBG pipe: process_block c04b17d3 at 23730 with 5 inputs and 11 outputs.
Dec 21 09:26:48.793 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.796 DEBG process_block: orphan: c04b17d3, # orphans 45
Dec 21 09:26:48.796 DEBG pipe: process_block 7b780e1a at 23732 with 1 inputs and 3 outputs.
Dec 21 09:26:48.796 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.799 DEBG process_block: orphan: 7b780e1a, # orphans 45
Dec 21 09:26:48.799 DEBG pipe: process_block 6e27f874 at 23734 with 4 inputs and 7 outputs.
Dec 21 09:26:48.799 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.802 DEBG process_block: orphan: 6e27f874, # orphans 45
Dec 21 09:26:48.803 DEBG pipe: process_block 527f285b at 23735 with 1 inputs and 3 outputs.
Dec 21 09:26:48.803 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.806 DEBG process_block: orphan: 527f285b, # orphans 45
Dec 21 09:26:48.806 DEBG pipe: process_block 59d68f34 at 23741 with 3 inputs and 7 outputs.
Dec 21 09:26:48.806 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.810 DEBG process_block: orphan: 59d68f34, # orphans 45
Dec 21 09:26:48.810 DEBG pipe: process_block d8a90fee at 23743 with 2 inputs and 5 outputs.
Dec 21 09:26:48.810 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.814 DEBG process_block: orphan: d8a90fee, # orphans 45
Dec 21 09:26:48.814 DEBG pipe: process_block bbd5e9bf at 23745 with 1 inputs and 3 outputs.
Dec 21 09:26:48.815 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.817 DEBG process_block: orphan: bbd5e9bf, # orphans 45
Dec 21 09:26:48.818 DEBG pipe: process_block eab1d444 at 23748 with 1 inputs and 3 outputs.
Dec 21 09:26:48.818 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.820 DEBG process_block: orphan: eab1d444, # orphans 45
Dec 21 09:26:48.821 DEBG pipe: process_block 324a4a9e at 23752 with 1 inputs and 3 outputs.
Dec 21 09:26:48.821 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.824 DEBG process_block: orphan: 324a4a9e, # orphans 45
Dec 21 09:26:48.824 DEBG pipe: process_block 4593ee04 at 23754 with 0 inputs and 1 outputs.
Dec 21 09:26:48.824 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.827 DEBG process_block: orphan: 4593ee04, # orphans 45
Dec 21 09:26:48.828 DEBG pipe: process_block 8f142e89 at 23757 with 1 inputs and 3 outputs.
Dec 21 09:26:48.828 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.831 DEBG process_block: orphan: 8f142e89, # orphans 45
Dec 21 09:26:48.831 DEBG pipe: process_block e1f23580 at 23738 with 1 inputs and 3 outputs.
Dec 21 09:26:48.832 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.834 DEBG process_block: orphan: e1f23580, # orphans 45
Dec 21 09:26:48.834 DEBG pipe: process_block fab36023 at 23762 with 7 inputs and 15 outputs.
Dec 21 09:26:48.835 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.838 DEBG process_block: orphan: fab36023, # orphans 45
Dec 21 09:26:48.838 DEBG pipe: process_block d9a3401a at 23725 with 3 inputs and 7 outputs.
Dec 21 09:26:48.838 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.841 DEBG process_block: orphan: d9a3401a, # orphans 45
Dec 21 09:26:48.841 DEBG pipe: process_block c0ae83a2 at 23764 with 0 inputs and 1 outputs.
Dec 21 09:26:48.841 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.845 DEBG process_block: orphan: c0ae83a2, # orphans 45
Dec 21 09:26:48.846 DEBG pipe: process_block 59c2698a at 23722 with 1 inputs and 3 outputs.
Dec 21 09:26:48.846 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.848 DEBG process_block: orphan: 59c2698a, # orphans 45
Dec 21 09:26:48.848 DEBG pipe: process_block 44fe463c at 23731 with 1 inputs and 3 outputs.
Dec 21 09:26:48.848 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.851 DEBG process_block: orphan: 44fe463c, # orphans 45
Dec 21 09:26:48.851 DEBG pipe: process_block 094a607b at 23737 with 0 inputs and 1 outputs.
Dec 21 09:26:48.851 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.853 DEBG process_block: orphan: 094a607b, # orphans 45
Dec 21 09:26:48.854 DEBG pipe: process_block b93b01c6 at 23727 with 5 inputs and 11 outputs.
Dec 21 09:26:48.854 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.857 DEBG process_block: orphan: b93b01c6, # orphans 45
Dec 21 09:26:48.857 DEBG pipe: process_block cb21a6a9 at 23740 with 1 inputs and 3 outputs.
Dec 21 09:26:48.857 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.860 DEBG process_block: orphan: cb21a6a9, # orphans 45
Dec 21 09:26:48.860 DEBG pipe: process_block 6f3fe086 at 23746 with 7 inputs and 15 outputs.
Dec 21 09:26:48.860 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.862 DEBG process_block: orphan: 6f3fe086, # orphans 45
Dec 21 09:26:48.863 DEBG pipe: process_block 201ab4ab at 23749 with 4 inputs and 9 outputs.
Dec 21 09:26:48.863 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.866 DEBG process_block: orphan: 201ab4ab, # orphans 45
Dec 21 09:26:48.866 DEBG pipe: process_block efed45f5 at 23753 with 0 inputs and 1 outputs.
Dec 21 09:26:48.866 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.869 DEBG process_block: orphan: efed45f5, # orphans 45
Dec 21 09:26:48.869 DEBG pipe: process_block 815c932d at 23736 with 1 inputs and 3 outputs.
Dec 21 09:26:48.869 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.872 DEBG process_block: orphan: 815c932d, # orphans 45
Dec 21 09:26:48.872 DEBG pipe: process_block 990410c8 at 23742 with 3 inputs and 7 outputs.
Dec 21 09:26:48.872 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.874 DEBG process_block: orphan: 990410c8, # orphans 45
Dec 21 09:26:48.875 DEBG pipe: process_block fe8ba00f at 23744 with 0 inputs and 1 outputs.
Dec 21 09:26:48.875 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.877 DEBG process_block: orphan: fe8ba00f, # orphans 45
Dec 21 09:26:48.877 DEBG pipe: process_block cc57c805 at 23751 with 0 inputs and 1 outputs.
Dec 21 09:26:48.877 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.880 DEBG process_block: orphan: cc57c805, # orphans 45
Dec 21 09:26:48.880 DEBG pipe: process_block 65951661 at 23756 with 1 inputs and 3 outputs.
Dec 21 09:26:48.881 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.884 DEBG process_block: orphan: 65951661, # orphans 45
Dec 21 09:26:48.884 DEBG pipe: process_block ad7f2579 at 23755 with 1 inputs and 3 outputs.
Dec 21 09:26:48.884 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.887 DEBG process_block: orphan: ad7f2579, # orphans 45
Dec 21 09:26:48.887 DEBG pipe: process_block 83ee3099 at 23758 with 1 inputs and 3 outputs.
Dec 21 09:26:48.888 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.890 DEBG process_block: orphan: 83ee3099, # orphans 45
Dec 21 09:26:48.890 DEBG pipe: process_block 2a6126a4 at 23759 with 1 inputs and 3 outputs.
Dec 21 09:26:48.890 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.894 DEBG process_block: orphan: 2a6126a4, # orphans 45
Dec 21 09:26:48.894 DEBG pipe: process_block d4cb8477 at 23747 with 5 inputs and 11 outputs.
Dec 21 09:26:48.895 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.900 DEBG process_block: orphan: d4cb8477, # orphans 45
Dec 21 09:26:48.900 DEBG pipe: process_block 9e318584 at 23760 with 7 inputs and 15 outputs.
Dec 21 09:26:48.900 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.903 DEBG process_block: orphan: 9e318584, # orphans 45
Dec 21 09:26:48.903 DEBG pipe: process_block eda5c219 at 23750 with 0 inputs and 1 outputs.
Dec 21 09:26:48.903 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.906 DEBG process_block: orphan: eda5c219, # orphans 45
Dec 21 09:26:48.906 DEBG pipe: process_block 84ef470d at 23763 with 469 inputs and 5 outputs.
Dec 21 09:26:48.906 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.909 DEBG process_block: orphan: 84ef470d, # orphans 45
Dec 21 09:26:48.909 DEBG pipe: process_block 02076f97 at 23720 with 2 inputs and 5 outputs.
Dec 21 09:26:48.909 DEBG pipe: validate_header cuckoo size 16
Dec 21 09:26:48.914 DEBG Starting new sumtree extension.
Dec 21 09:26:49.124 DEBG pipe: proces_block 02076f97 at 23720 is valid, save and append.
Dec 21 09:26:49.125 INFO pipe: update_head: 02076f97, 177871453 at 23720
Dec 21 09:26:49.126 DEBG Committing sumtree extension.
Dec 21 09:26:49.396 DEBG Sumtree extension done.
```

**Comment by antiochp** [comment]
Resolved in #537
